nixos/modules/services/misc/rmfakecloud.nix

```diff
@@ -0,0 +1,119 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.rmfakecloud;
+  serviceDataDir = "/var/lib/rmfakecloud";
+
+in {
+  options = {
+    services.rmfakecloud = {
+      enable = mkEnableOption "rmfakecloud remarkable self-hosted cloud";
+
+      storageUrl = mkOption {
+        type = types.str;
+        example = "https://local.appspot.com";
+        description = ''
+          Controls whether file upload/download goes through the local proxy
+          or to an external server. It's the address of rmfakecloud as visible
+          from the tablet, especially if the host is behind a reverse proxy or
+          in a container.
+        '';
+      };
+
+      port = mkOption {
+        type = types.port;
+        default = 3000;
+        description = ''
+          Listening port number.
+        '';
+      };
+
+      logLevel = mkOption {
+        type = types.enum [ "info" "debug" "warn" "error" ];
+        default = "info";
+        description = ''
+          Set the log verbosity.
+        '';
+      };
+
+      extraSettings = mkOption {
+        type = with types; attrsOf str;
+        default = { };
+        example = { RM_HTTPS_COOKIE = true; };
+        description = ''
+          Extra settings in the form of a set of key-value pairs.
+          For tokens and secrets, use `environmentFile` instead.
+        '';
+      };
+
+      environmentFile = mkOption {
+        type = with types; nullOr path;
+        default = null;
+        example = "/etc/secrets/rmfakecloud.env";
+        description = ''
+          Path to an environment file.
+          This can be used to securely store tokens and secrets outside of the
+          world-readable Nix store.
+        '';
+      };
+    };
+  };
+
+  config = mkIf cfg.enable {
+    systemd.services.rmfakecloud = {
+      description = "rmfakecloud remarkable self-hosted cloud";
+
+      environment = {
+        STORAGE_URL = cfg.storageUrl;
+        PORT = toString cfg.port;
+        LOGLEVEL = cfg.logLevel;
+      } // cfg.extraSettings;
+
+      preStart = ''
+        # Generate the secret key used to sign client session tokens.
+        # Replacing it invalidates the previously established sessions.
+        if [ ! -f jwt_secret_key ]; then
+          (umask 077; touch jwt_secret_key)
+          ${pkgs.openssl}/bin/openssl rand -base64 48 >> jwt_secret_key
```

Comment:
I'm a bit troubled by this since `JWT_SECRET_KEY` may have been defined in the environment file, thus making this code not only rather unuseful but also confusing by creating an unused secret.

Moreover, the extra dependency on openssl pushes an additional 90MB to the closure (it could be argued that openssl may already be in the system, but I'm still reticent to add it here).

I see two acceptable solutions:

- Add an option `autoCreateJwtSecret`, defaults to true and informs the user that if such secret does not exist and is not defined in the envvars, it shall be created. If set to false, don't bring openssl into the closure but merely assert that either the file exists or the envvar is defined.
- Check whether the envvar is defined before creating the secret. I'm not sure whether preStart has access to the environment (my guess it that it does).

Reply:
Agreed. I implemented your second suggestion. I also removed the dependency over `openssl` for generating the random secret.